### PR TITLE
Fix: Set correct branch when it's not specified in the config

### LINKED ISSUE #5617: Default to branch name `main` rather than `master`
**Describe the bug**
The documents reference a default branch `master`; this should be changed to `main` if possible, especially seeing as the Netlify deployment documents seem to encourage that too, and github's changing its default to `main` as well.

**To Reproduce**
1. https://www.netlifycms.org/docs/add-to-your-site/#backend
2. See comment `branch: master # Branch to update (optional; defaults to master)`

**Expected behavior**
I'd expect the branch to default to `main`, or at least the documentation to acknowledge that's becoming the new default in most places.

### PR DESCRIPTION
**Pull request by bytrangle**
fix #5617 

**Summary**
Currently, the default branch in Github, Gitlab and Bitbucket is set to be `main` instead of `master`.

However, when people use Netlify CMS and don't specify the branch name to push content in the config file, Netlify CMS will automatically push to branch `master`.

Since this branch may not exist, users will get an error on the frontend: "**API_ERROR: Branch not found**". 

**Test plan**
1. In the `admin/config.yml` file, specify a valid Github repo path to push content to.
2. Run `yarn start`.
3. Go to Neltify CMS dashboard, create a new post and hit Publish. You shouldn't get an error. Checking back on the repo that you specified in the config, you should see the new post.

**Checklist**

Please add a `x` inside each checkbox:

- [x] I have read the [contribution guidelines](../CONTRIBUTING.md).
- [x] Code is formatted via running `yarn format`.
- [ ] Tests are passing via running `yarn test`.
- [ ] The status checks are successful (continuous integration). Those can be seen below.

### CODE REVIEW

**Comment by bytrangle** [comment]
Please hold off on reviewing until I've pushed the unit tests.

**Review by erezrokah — CHANGES_REQUESTED** [review]
Thanks @bytrangle, great work as always.

I've added some comments, please let me know your thoughts.
Also, are we planning to do this for other backends too?

File: packages/netlify-cms-backend-github/src/implementation.tsx
Comment (bytrangle) [reply packages/netlify-cms-backend-github/src/implementation.tsx:189]: Three things:
1. I didn't consider the scenario that `authenticate` can be revoked. I will change it as requested.Good point.

2. I was going to apply the same approach to other backend, but then I realized that would mean writing some repetitive code. That will open up more room for errors and make it harder to track changes in the future. So what do you think about adding a helper in Netlify Utils to get a default branch and call that in each respective backend?

Comment (erezrokah) [reply packages/netlify-cms-backend-github/src/implementation.tsx:189]: > 2\. Netlify Utils to get a default branch and call that in each respective backend?

Sounds good! How about adding the helpers to https://github.com/netlify/netlify-cms/blob/master/packages/netlify-cms-lib-util/src/backendUtil.ts?

Comment (bytrangle) [reply packages/netlify-cms-backend-github/src/implementation.tsx:189]: Should it be backendUtils or  [API.ts](https://github.com/netlify/netlify-cms/blob/master/packages/netlify-cms-lib-util/src/API.ts) is the place?
And when do you need these changes? I normally don't work on open source on weekend, but if you need it soon, I can work on this issue on Sunday.

Comment (erezrokah) [reply packages/netlify-cms-backend-github/src/implementation.tsx:189]: Either `backendUtils` or `API.ts` work (or both) depending on the common logic you extract.

> I normally don't work on open source on weekend, but if you need it soon, I can work on this issue on Sunday.

I definitely don't need it soon, and in any case no need to change your work schedule for us.

**Comment by bytrangle** [comment]
@erezrokah I'm so sorry for dropping the ball on this. Do you still need this issue to be worked on? I can get back to it in a few days.

**Comment by erezrokah** [comment]
>@erezrokah I'm so sorry for dropping the ball on this. Do you still need this issue to be worked on? I can get back to it in a few days.

No worries, I appreciate the work you've done so far. I'd be happy to see this land and merged if you have the time to finalize it.

**Comment by bytrangle** [comment]
@erezrokah Thank you for your kind words.

**Update**: My changes have led to a bunch of failed unit tests. In my newest [commit](https://github.com/netlify/netlify-cms/pull/5844/commits/346ac8ba710a5c174bc5d03003d2ab78681e53d6), I tackled the first failed test I saw named [stores and returns user object on success](https://github.com/netlify/netlify-cms/blob/master/packages/netlify-cms-backend-gitlab/src/__tests__/gitlab.spec.js#L339). 

What happens under the hood is the test calls an `authenticate` method. This method calls the endpoint `/projects/:id` **twice**:
- the first time for checking if the authenticated user has write access.
- the second time is for getting the default branch name.

However, the test only intercepts the request to `projects/:id` **once** through the `interceptAuth()` function. 

```
 function interceptAuth(backend, { userResponse, projectResponse } = {}) {
    const api = mockApi(backend);
    // ...
    api
      .get(expectedRepoUrl)
      .query(true)
      .reply(200, projectResponse || resp.project.success);
  }
```
As a result, the second API request from my codes is never intercepted by Nock, leading to an error: "Nock: No match for request ...  "url": "https://gitlab.com/api/v4/projects/foo%2Fbar"".

Hope that makes sense.

**Comment by bytrangle** [comment]
@erezrokah I'm trying to test if the function `getDefaultBranchName` is called when the config for each backend misses the value for `branch` field.

Here's what I'm unsure about where to start: 
- I added `getDefaultBranchName` function to `netlify-cms-lib-util` so that each backend can import it, thus avoiding repetition.
- However, the function that calls `getDefaultBranchName` is `authenticate` method defined as a property of each namesake class.

So where should I put the test for checking if `getDefaultBranchName` is called? In the tests folder for each backend, or in the test folder of `netlify-cms-lib-util`?

For now, I'm putting the test in `netlify-cms-backend-gitlab/src/__tests/gitlab.spec.js` and it passed.

```
// ...
jest.mock('netlify-cms-lib-util', () => ({
  ...jest.requireActual('netlify-cms-lib-util'),
  getDefaultBranchName: jest.fn().mockResolvedValue('master')
}))
// ...
    it('getDefaultBranchName is called when branch field value is empty in the config', async () => {
      const backendName = defaultConfig.backend.name
      backend = resolveBackend(defaultConfig)
      interceptAuth(backend)
      await backend.authenticate(mockCredentials)
      expect(1).toEqual(1)
      expect(getDefaultBranchName).toHaveBeenCalledTimes(1)
      expect(getDefaultBranchName).toHaveBeenCalledWith({ backend: backendName, repo, token: mockCredentials.token })
    })
```

What do you think of it?

**Comment by bytrangle** [comment]
@erezrokah Would you mind review this PR again since I've made significant changes to it since the last review?

**Comment by bytrangle** [comment]
@erezrokah Thank you for the reformat work. I was doing it but you're faster.

**Review by erezrokah — CHANGES_REQUESTED** [review]
Thanks for following up with these changes @bytrangle, I added some comments.

Also, in order for this change to be backwards compatible we would need to default to `master` if exists (default or not), and only if it doesn't exist use the default branch.

Please let me know what you think

File: packages/netlify-cms-backend-bitbucket/src/implementation.ts
Comment (bytrangle) [reply packages/netlify-cms-backend-bitbucket/src/implementation.ts:229]: Indeed, I kept wrestling with the question of when to set `this.branch`. I put it after the check for user having write access because I thought there was no point setting branch when user is unauthenticated or doesn't have write access.

What do you recommend? Setting branch before calling `hasWriteAccess`, or even earlier than `authenticate`?

File: packages/netlify-cms-lib-util/src/API.ts
Comment (bytrangle) [reply packages/netlify-cms-lib-util/src/API.ts:219]: Yes, that's a good point. I think I was doing that (duplicate `getDefaultBranchName` per backend) in the draft. Then I thought maybe I would make it a little more efficient by putting it in the lib. But it wasn't.

If we are to have `getDefaultBranchName` in each backend API, should it be in the `implementation` module?

**Comment by martinjagodic** [comment]
@bytrangle are you still interested in moving this forward?

**Comment by paulRbr** [comment]
I think this PR is interesting to be merged upstream. I was going to suggest [a _much more simple_ change](https://github.com/paulRbr/decap-cms/commit/8da23d2ceeb05af4cdc9b890cd1e9b97cc988722) (changing the default value of the branch to `main`) but I think this PR is better.

**Comment by martinjagodic** [comment]
@paulRbr yes, we would like to have this merged, but we need to solve those merge conflicts first.

**Comment by bytrangle** [comment]
@martinjagodic Hi, I'm interested in working on this again. However, I may have to ask you about resolving merging conflicts later :).

**Comment by martinjagodic** [comment]
@bytrangle awesome! The conflicts emerged when we renamed the packages from `netlify-cms-*` to `decap-cms-*`. If a rename does not suffice, I am happy to answer questions/help out.

**Comment by bytrangle** [comment]
Hi @demshy . Is there anything else I need to do?

**Comment by demshy** [comment]
Sorry @bytrangle I somehow overlooked this one. I added a missing import that made the unit tests fail. I will gladly merge this one into a release this week

**Review by demshy — APPROVED** [review]

File: packages/netlify-cms-lib-util/src/API.ts
Comment (floscher) [inline packages/netlify-cms-lib-util/src/API.ts]: Hi @bytrangle, do you by any chance remember, why this is using `token` instead of bearer here?
This seems to be causing an issue with Gitlab logins, since this PR was merged in April: https://github.com/decaporg/decap-cms/issues/7172#issuecomment-2184138768

I'm not sure about Github/Bitbucket, but it seems at least for Gitlab this line should be like this:

```suggestion
    baseHeaders['Authorization'] = `Bearer ${token}`;
```

---

For reference:  This line was originally introduced in 3edc3cef1eeb905af25010e50524925c2748603e, this comment is on a2add4a4954aafa420f217af1d68867cc786e33a, which was later squashed to c91a70fbea7d6a4d413fc00cdd4a49b24fa97a16

Comment (bytrangle) [reply packages/netlify-cms-lib-util/src/API.ts]: I don't remember. This corresponds to Gitlab docs 13.0, which is no longer available. Please feel free to make a new PR to correct it.

Comment (floscher) [reply packages/netlify-cms-lib-util/src/API.ts]: Thanks anyway, I now created #7242 for this change.

Comment (b-xb) [reply packages/netlify-cms-lib-util/src/API.ts]: Nice work @floscher

I cloned the decapCMS repo locally yesterday to double the fix.

After adding the fix and rebuilding decapCMS, I no longer got the previous errir message.

However, I then still got an "API_ERROR: Branch not found" message, implying that the code this patch was meant to fix doesn't do it's job on gitlab anyway..

Only after hard-coding "branch: main" into my config.yml, was I then able to get on with things.

**Pull request merged**
